shared/slackutil/messaging.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// messaging.go includes functions to send message to Slack channel.
+
+package slackutil
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"net/http"
+	"net/url"
+)
+
+const postMessageURL = "https://slack.com/api/chat.postMessage"
+
+// Operations defines the operations that can be done to Slack
+type Operations interface {
+	Post(text, channel string) error
+}
+
+// client contains Slack bot related information
+type client struct {
```

Comment:
I think this needs to be exposed

Reply:
It doesn't have to, see `Interface factories` in https://www.sohamkamani.com/blog/golang/2018-06-20-golang-factory-patterns/.

The benefit of writing like this is, we do not expose it to the user, so user will only be able to construct the `client` with `NewClient`, in which we can enforce validations for the input params.

Reply:
I see, that's good to learn, the downside I can think about this approach is user will not be able to cast, but it's ok in this case, since it's just a simple wrapper